### PR TITLE
⏬ #2265 Entity Data table sort

### PR DESCRIPTION
**Pull request by demariadaniel**
# Description of changes

Handles files w/ errors + sorting by core completion columns

https://app.zenhub.com/workspaces/icgc-argo-platform-dk-production-board-5e542d38415f5034e9fed89d/issues/icgc-argo/platform-ui/2265

## Type of Change

- [ ] Bug
- [x] Styling
- [ ] New Feature

## Checklist before requesting review

- Design (select one):
  - [ ] Matches design:
    - component sizes, spacing, and styles
    - font size, weight, colour
    - spelling has been double checked
  - [x] No design provided, screenshot of changes included in PR
  - [ ] No changes to UI
- [x] Feature is minimally responsive
- [x] Manual testing of UI feature
- [x] Add copyrights to new files
- [x] Connected ticket to PR

### CODE REVIEW

**Comment by samrichca** [comment]
just checking, the API data for this table arrives all at once? it's not sorted or paginated?

**Review by ciaranschutte — COMMENTED** [review]
couple of improvement suggestion / food for thought. will take another look in the morning

File: components/pages/submission-system/program-submitted-data/ClinicalEntityDataTable/index.tsx [hunk 1]
```diff
@@ -283,6 +284,33 @@ const ClinicalEntityDataTable = ({
...
 
+  const sortEntityData = (prev, next) => {
+    let sortVal = 0;
+
+    if (hasErrors) {
+      // If Current Entity has Errors, Prioritize Data w/ Errors
+      const errorsA = clinicalErrors.find((error) => error.donorId == prev['donor_id']) ? -1 : 0;
+      const errorsB = clinicalErrors.find((error) => error.donorId == next['donor_id']) ? 1 : 0;
+      sortVal += errorsA + errorsB;
+    }
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply components/pages/submission-system/program-submitted-data/ClinicalEntityDataTable/index.tsx:295]: completionSortIndex is meant to handle the user manually clicking on a Completion column header and sorting by that key
It is intended to override sort by errors b/c the user is manually sorting by a specific field

File: components/pages/submission-system/program-submitted-data/ClinicalEntityDataTable/index.tsx [hunk 2]
```diff
@@ -283,6 +284,33 @@ const ClinicalEntityDataTable = ({
...
+
+    // Handles Sorting by Core Completion columns
+    const completionColumns = Object.values(aliasSortNames);
+    const completionSortIndex = completionColumns.indexOf(sortKey);
+
+    if (completionSortIndex) {
+      const columnNames = Object.keys(aliasSortNames);
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply components/pages/submission-system/program-submitted-data/ClinicalEntityDataTable/index.tsx:302]: good call, wasn't thinking about re-renders, just struggling to keep up w/ the proliferation of utility variables to keep track of the many aliases and names, i.e. `primaryDiagnosis = 'primaryDiagnosis' = primaryDiagnoses = PD = primary_diagnoses = Primary Diagnosis` ;)

File: components/pages/submission-system/program-submitted-data/ClinicalEntityDataTable/index.tsx [hunk 3]
```diff
@@ -503,6 +531,7 @@ const ClinicalEntityDataTable = ({
         page={page}
         pages={numTablePages}
         pageSize={pageSize}
+        defaultSortMethod={sortEntityData}
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply components/pages/submission-system/program-submitted-data/ClinicalEntityDataTable/index.tsx:534]: Yes it's this way for a reason but maybe there's improvements to be made

So by default we want errors prioritized, followed by clinical completion, followed by donorID.
Entities are already sorted by `donorId`, `coreCompletionPercentage` + requested sort key on the service side.

However `Donor Entity` data and `Donor Error` data are separate queries + service endpoints, so the front end is just harmonizing the two requests here, it's just handling the error part.

Then if the user sorts by specific column that overrides the default.

defaultSortMethod is recommended use for custom sorting logic
https://www.npmjs.com/package/react-table/v/6.10.0#custom-sorting-algorithm

Service PR
https://github.com/icgc-argo/argo-clinical/pull/796

Caching seems to be happening based on service logging/network tab
But could follow up if you see any potential there!

File: components/pages/submission-system/program-submitted-data/ClinicalEntityDataTable/index.tsx [hunk 4]
```diff
@@ -283,6 +284,33 @@ const ClinicalEntityDataTable = ({
     };
   });
 
+  const sortEntityData = (prev, next) => {
+    let sortVal = 0;
+
+    if (hasErrors) {
+      // If Current Entity has Errors, Prioritize Data w/ Errors
+      const errorsA = clinicalErrors.find((error) => error.donorId == prev['donor_id']) ? -1 : 0;
+      const errorsB = clinicalErrors.find((error) => error.donorId == next['donor_id']) ? 1 : 0;
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply components/pages/submission-system/program-submitted-data/ClinicalEntityDataTable/index.tsx:293]: Thanks for the reduce! I was specifically using `==` b/c Entity donorId is a string + Error donorId is a number
Will see if I can track this down + update on the back end so we can use `===`

**Comment by demariadaniel** [comment]
> just checking, the API data for this table arrives all at once? it's not sorted or paginated?

@samrichca the data _is_ sorted + paginated on the back end but there are discrepancies between Clinical Entities + Clinical Errors, as well as coordination between React Table + Mongoose sort/pagination which needs to happen

Looking into these specifically in review!

File: components/pages/submission-system/program-submitted-data/ClinicalEntityDataTable/index.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (demariadaniel) [reply components/pages/submission-system/program-submitted-data/ClinicalEntityDataTable/index.tsx]: Updated

File: components/pages/submission-system/program-submitted-data/ClinicalEntityDataTable/index.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (demariadaniel) [reply components/pages/submission-system/program-submitted-data/ClinicalEntityDataTable/index.tsx:527]: So I ended up looking into this and doing a significant refactor

I found a bug in the API which was messing up sort-by Clinical Completion. With that fixed I am able to keep Completion sort logic on the service and remove logic from the FE.

Error Sort I was able to move to the API, however there's some tradeoffs with the refactor. I moved discussion to that branch. There's an argument to be made for keeping Error Sort on the front end.

Any feedback on better ways to organize this, API PR is here: https://github.com/icgc-argo/platform-api/pull/585/files

[the remark this answers is not in the document]
Comment (demariadaniel) [reply components/pages/submission-system/program-submitted-data/ClinicalEntityDataTable/index.tsx:545]: I've reverted to Error sorting on the FE until we align on approach for API solution

**Review by samrichca — APPROVED** [review]
Looks good, feedback was addressed & discussed offline as well. lgtm!

**Pull request merged**